src/lib/branchpredictors/TagePredictor.cc

```diff
@@ -0,0 +1,283 @@
+    } else {
+      // If history entry belongs to a branch-and-link instruction, pop target
+      // off of stack
+      if (ras_.size()) {
+        ras_.pop_back();
+      }
+    }
+    rasHistory_.erase(it);
+  }
+
+  assert((ftq_.size() > 0) &&
+         "Cannot flush instruction from Branch Predictor "
+         "when the ftq is empty");
+  ftq_.pop_back();
+
+  // Roll back global history
+  globalHistory_.rollBack();
+}
+
+void TagePredictor::getTaggedPrediction(uint64_t address,
+                                        BranchPrediction* prediction,
+                                        BranchPrediction* altPrediction,
+                                        int8_t* predTable,
+                                        std::shared_ptr<uint64_t[]> indices,
+                                        std::shared_ptr<uint64_t[]> tags) {
+  // Get a basic prediction from the btb
+  BranchPrediction basePrediction = getBtbPrediction(address);
+  prediction->isTaken = basePrediction.isTaken;
+  prediction->target = basePrediction.target;
+  *predTable = -1;
+
+  // Check each of the tagged predictor tables for an entry matching this
+  // branch.  If found, update the best prediction.  The greater the table
+  // number, the longer global history it has access to.  Therefore, the
+  // greater the table number, the better the prediction.
+  for (int8_t table = 0; table < numTageTables_; table++) {
+    // Determine the index and tag for this table, as they vary depending on
+    // the length of global history
+    uint64_t index = getTaggedIndex(address, table);
+    indices.get()[table] = index;
+    uint64_t tag = getTag(address, table);
+    tags.get()[table] = tag;
+
+    // If tag matches, then use this prediction
+    if (tageTables_[table][index].tag == tag) {
+      altPrediction->isTaken = prediction->isTaken;
+      altPrediction->target = prediction->target;
+
+      prediction->isTaken = (tageTables_[table][index].satCnt >= 2);
+      prediction->target = tageTables_[table][index].target;
+      *predTable = table;
+    }
+  }
+}
+
+BranchPrediction TagePredictor::getBtbPrediction(uint64_t address) {
+  // Get prediction from BTB
+  uint64_t index = (address >> 2) & ((1ull << btbBits_) - 1);
+  bool direction = (btb_[index].first >= (1 << (satCntBits_ - 1)));
+  uint64_t target = btb_[index].second;
+  return {direction, target};
+}
+
+uint64_t TagePredictor::getTaggedIndex(uint64_t address, uint8_t table) {
+  // Get the XOR of the address (sans two least-significant bits) and the
+  // global history (folded onto itself to make it of the correct size).
+  uint64_t h1 = (address >> 2);
+  uint64_t h2 = globalHistory_.getFolded(1ull << (table + 1), tageTableBits_);
+  // Then truncate the XOR to make it fit the desired size of an index
+  return (h1 ^ h2) & ((1 << tageTableBits_) - 1);
+}
+
+uint64_t TagePredictor::getTag(uint64_t address, uint8_t table) {
+  // Hash function here is pretty arbitrary
+  uint64_t h1 = address;
```

Comment:
Any reason not to remove the 2 LSBs here?

Reply:
Yes.  Ideally the hashes for the tag and the index should never each produce the same value for two different branches.  Therefore, because the index does remove the 2 LSBs, keeping them here makes the information being passed into the hashes different and so improves the accuracy of the BP (reduces the risk of this type of accidental clashing).